crates/noirc_frontend/src/parser/mod.rs

```diff
@@ -181,7 +182,7 @@ fn parameter_name_recovery<T: Recoverable + Clone>() -> impl NoirParser<T> {
 fn top_level_statement_recovery() -> impl NoirParser<TopLevelStatement> {
     none_of([Token::Semicolon, Token::RightBrace, Token::EOF])
         .repeated()
-        .ignore_then(one_of([Token::Semicolon, Token::RightBrace]))
+        .ignore_then(one_of([Token::Semicolon]))
```

Comment:
Why are we getting rid of `Token::RightBrace` here with the addition of parsing submodules as a top-level statement, don't we want the recovery to still finish with a right brace for submodules (as well as the other top-level statements like struct)?

Reply:
Good question :)

This is a subtle one. This is the recovery strategy for what to do when a top-level statement fails to parse. What we used to do was skip to the first semicolon, rightbrace, or EOF and consume it to recover. This worked fine before (though could sometimes skip to a non-matching } or ; e.g. within a function and cause us to emit more errors than needed).

With submodules added however, we can have valid files such as:
```
mod foo {
    fn bar() { }
    fn baz() { }
}
```
Where inside the submodule we also parse a sequence of top-level-statements. The first two of which correctly parse as functions, then it tries to parse a third and fails. When it fails it runs the recovery parser which skips _and passes_ the terminating } that the mod is expecting. So when this then ends with the statements { fn, fn, error } after, the submodule rule expects `just(Token::RightBrace)` and fails because it was already consumed. So the error recovery strategy had to be changed.

Ideally in the future we should change it so it should only attempt to recover if at least one token was successfully parsed. This may be more difficult than it sounds though.

Reply:
Ok that makes sense thank you for the explanation. Your last statement seems like a worthy goal as well, but probably within a separate PR. The change looks like it does not break any of the other top level statement parsing as well so the PR LGTM.